{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "126e17c8",
   "metadata": {},
   "source": [
    "# Compare Old and New STRait Razor Results\n",
    "\n",
    "We first tried using STRait Razor with .config files that were generated directly from PERF that was run on the reference .fasta sequence for all thirty primers. Each .config file corresponded to a single primer and contained one 5' and 3' flanking region pair for the STR. \n",
    "\n",
    "We found that running STRait Razor resulted in incorrect genotypes when indels were present (Primer 25) - this was due to the .config file not containing flanking region options for samples with and without the indel. We also found that results were only available for 159 out of 196 possible genotypes, but for these 159 primer/sample pairs, STRait Razor had 75% accuracy.\n",
    "\n",
    "To make .config files that are robust to indels, we ran PERF on every single read for all primer/sample pairings, and then chose the most frequent flanking regions for our .config files (a flanking region was deemed \"frequent\" if it contained > 7.5% of all reads, forward and reverse, since we chose 15% as our allele identification threshold earlier). (***this threshold needs to be fixed***). We reasoned that including all reads would generate .config files that contained not just one haplotype, but many haplotypes for the flanking regions, allowing all samples to be \"matched\" to flanking regions and then genotyped by STRait Razor.\n",
    "\n",
    "Our .config files on average tended to have more than one line - many lines were repeats with only one base pair off, but others contained the repeat motif and were \"false\" flanking regions. However, many lines did account for indels. Nonetheless, these .config files need to be fixed with a \"sliding window\" approach to process false flanking regions, but before doing so, we need to identify which primers and samples were the most problematic for STRait razor run using the new .config files, which had ~60% accuracy, but for 172 primer/sample pairs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3d971131",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b6b0637",
   "metadata": {},
   "source": [
    "## See which samples failed before/after new .config file creation\n",
    "\"Failed\" in this case means that the STRait razor output file was empty, not that the genotyping was incorrect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54494136",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Primers in Excel Table:\n",
      "**********************************************\n",
      "[ 3  4  5  8 10 11 12 13 14 15 16 17 18 19 20 22 23 25 26 28 29 30]\n",
      "Primers genotyped using old .config files:\n",
      "**********************************************\n",
      "[ 3  4  8 10 11 12 13 14 15 16 17 18 20 22 23 25 26 28 29 30]\n",
      "Primers genotyped using new .config files:\n",
      "**********************************************\n",
      "[ 3  4  5  8 10 11 12 13 14 15 16 17 18 20 22 23 25 26 28 29 30]\n",
      "\n",
      "Failed pairings after new .config files:\n",
      "*****************************************************\n",
      "20         (5, 1232)\n",
      "21         (5, 1393)\n",
      "29        (5, 34966)\n",
      "81     (14, 2006174)\n",
      "85        (15, 1393)\n",
      "89         (15, 217)\n",
      "91     (15, 2599208)\n",
      "110       (18, 1232)\n",
      "117       (19, 1232)\n",
      "118       (19, 1393)\n",
      "119    (19, 2006092)\n",
      "120        (19, 217)\n",
      "121    (19, 2520669)\n",
      "122    (19, 2599208)\n",
      "123      (19, 34896)\n",
      "124      (19, 34966)\n",
      "167    (26, 2006092)\n",
      "168        (26, 217)\n",
      "184    (29, 2006174)\n",
      "186      (29, 34966)\n",
      "191    (30, 2006174)\n",
      "192        (30, 217)\n",
      "194    (30, 2599208)\n",
      "195      (30, 34966)\n",
      "Name: P/S Pair, dtype: object\n",
      "\n",
      "Failed pairings after old .config files:\n",
      "*****************************************************\n",
      "20         (5, 1232)\n",
      "21         (5, 1393)\n",
      "22         (5, 1791)\n",
      "23      (5, 2006092)\n",
      "24      (5, 2006174)\n",
      "25          (5, 217)\n",
      "26      (5, 2520669)\n",
      "27      (5, 2599208)\n",
      "28        (5, 34896)\n",
      "29        (5, 34966)\n",
      "30         (8, 1232)\n",
      "33      (8, 2006092)\n",
      "34      (8, 2006174)\n",
      "36      (8, 2520669)\n",
      "38        (8, 34896)\n",
      "39        (8, 34966)\n",
      "81     (14, 2006174)\n",
      "85        (15, 1393)\n",
      "89         (15, 217)\n",
      "91     (15, 2599208)\n",
      "110       (18, 1232)\n",
      "117       (19, 1232)\n",
      "118       (19, 1393)\n",
      "119    (19, 2006092)\n",
      "120        (19, 217)\n",
      "121    (19, 2520669)\n",
      "122    (19, 2599208)\n",
      "123      (19, 34896)\n",
      "124      (19, 34966)\n",
      "167    (26, 2006092)\n",
      "168        (26, 217)\n",
      "184    (29, 2006174)\n",
      "186      (29, 34966)\n",
      "191    (30, 2006174)\n",
      "192        (30, 217)\n",
      "194    (30, 2599208)\n",
      "195      (30, 34966)\n",
      "Name: P/S Pair, dtype: object\n",
      "37\n",
      "\n",
      "Pairings that originally failed prior to new .config files, but worked after:\n",
      "22       (5, 1791)\n",
      "23    (5, 2006092)\n",
      "24    (5, 2006174)\n",
      "25        (5, 217)\n",
      "26    (5, 2520669)\n",
      "27    (5, 2599208)\n",
      "28      (5, 34896)\n",
      "30       (8, 1232)\n",
      "33    (8, 2006092)\n",
      "34    (8, 2006174)\n",
      "36    (8, 2520669)\n",
      "38      (8, 34896)\n",
      "39      (8, 34966)\n",
      "Name: P/S Pair, dtype: object\n"
     ]
    }
   ],
   "source": [
    "#Load in all excel table genotypes\n",
    "excel_pyrhulla_genotypes = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/\" +\\\n",
    "                                       \"excel_pyrhulla_genotypes.tsv\",sep=\"\\t\")\n",
    "\n",
    "#Load in all STRait razor genotypes using multiple flanks (new config files) or single flanks (old config files)\n",
    "full_merged_multiple_flanks_tsv = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/STRaitRazor/results/\" +\\\n",
    "                             \"alleles/full_merged_STRait_excel_alleles_0.15_multiple_flanks.tsv\",sep=\"\\t\")\n",
    "pd.options.display.max_colwidth = 100\n",
    "full_merged_multiple_flanks_tsv = full_merged_multiple_flanks_tsv.reindex(sorted \\\n",
    "                                        (full_merged_multiple_flanks_tsv.columns), axis=1)\n",
    "full_merged_single_flanks_tsv = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/STRaitRazor/results/\" +\\\n",
    "                             \"alleles/full_merged_STRait_excel_alleles_0.15_flank10.tsv\",sep=\"\\t\")\n",
    "full_merged_single_flanks_tsv = full_merged_single_flanks_tsv.reindex(sorted \\\n",
    "                                        (full_merged_single_flanks_tsv.columns), axis=1)\n",
    "\n",
    "#See if any primers are discarded by STRait Razor in either analysis\n",
    "print(\"Primers in Excel Table:\")\n",
    "print(\"**********************************************\")\n",
    "print(np.unique(excel_pyrhulla_genotypes[[\"Primer\"]]))\n",
    "\n",
    "print(\"Primers genotyped using old .config files:\")\n",
    "print(\"**********************************************\")\n",
    "print(np.unique(full_merged_single_flanks_tsv[[\"Primer\"]]))\n",
    "    \n",
    "print(\"Primers genotyped using new .config files:\")\n",
    "print(\"**********************************************\")\n",
    "print(np.unique(full_merged_multiple_flanks_tsv[[\"Primer\"]]))\n",
    "print(\"\")\n",
    "#See which primer/sample pairs from the excel table are discarded by STRait Razor in either analysis\n",
    "\n",
    "excel_pyrhulla_genotypes[\"P/S Pair\"]=excel_pyrhulla_genotypes[[\"Primer\",\"Sample\"]].apply(tuple,axis=1)\n",
    "full_merged_multiple_flanks_tsv[\"P/S Pair\"]=full_merged_multiple_flanks_tsv[[\"Primer\",\"Sample\"]].apply(tuple,axis=1)\n",
    "full_merged_single_flanks_tsv[\"P/S Pair\"]=full_merged_single_flanks_tsv[[\"Primer\",\"Sample\"]].apply(tuple,axis=1)\n",
    "\n",
    "print(\"Failed pairings after new .config files:\")\n",
    "print(\"*****************************************************\")\n",
    "multiple_flanks_failed_pairings=excel_pyrhulla_genotypes[\"P/S Pair\"][~excel_pyrhulla_genotypes[\"P/S Pair\"] \\\n",
    "                   .isin(full_merged_multiple_flanks_tsv[\"P/S Pair\"])]\n",
    "print(multiple_flanks_failed_pairings)\n",
    "print(\"\")\n",
    "\n",
    "print(\"Failed pairings after old .config files:\")\n",
    "print(\"*****************************************************\")\n",
    "single_flanks_failed_pairings=excel_pyrhulla_genotypes[\"P/S Pair\"][~excel_pyrhulla_genotypes[\"P/S Pair\"] \\\n",
    "                   .isin(full_merged_single_flanks_tsv[\"P/S Pair\"])]\n",
    "print(single_flanks_failed_pairings)\n",
    "print(len(single_flanks_failed_pairings))\n",
    "print(\"\")\n",
    "\n",
    "print(\"Pairings that originally failed prior to new .config files, but worked after:\")\n",
    "new_config_new_genotypes=single_flanks_failed_pairings[~single_flanks_failed_pairings.isin(multiple_flanks_failed_pairings)]\n",
    "print(new_config_new_genotypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ca89546",
   "metadata": {},
   "source": [
    "## Incorrect Genotyping Comparison\n",
    "* See which pairs \"failed\" (empty results) with old .config files but were wrongly genotyped with new .config files\n",
    "* See which pairs \"failed\" (empty results) with old .config files but were correctly genotyped with new .config files\n",
    "* See which pairs were correctly genotyped with old .config files but not with new .config files\n",
    "* See which pairs were correctly genotyped with new .config files but not with old .config files\n",
    "*etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a3ff2c67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of pairs that didn't run with old .config files but ran with new .config files:\n",
      "*****************************************************************\n",
      "13\n",
      "\n",
      "Pairs that didn't run with old .config files and were incorrectly genotyped with new .config files:\n",
      "*****************************************************************\n",
      "1         (5, 217)\n",
      "6        (5, 1791)\n",
      "9       (5, 34896)\n",
      "13    (5, 2006092)\n",
      "16    (5, 2006174)\n",
      "19    (5, 2520669)\n",
      "20    (8, 2520669)\n",
      "23    (5, 2599208)\n",
      "Name: P/S Pair, dtype: object\n",
      "Length:\n",
      "8\n",
      "Pairs that didn't run with old .config files and were correctly genotyped with new .config files:\n",
      "*****************************************************************\n",
      "30       (8, 1232)\n",
      "33    (8, 2006092)\n",
      "34    (8, 2006174)\n",
      "38      (8, 34896)\n",
      "39      (8, 34966)\n",
      "Name: P/S Pair, dtype: object\n",
      "Length:\n",
      "5\n",
      "Pairs that were correctly genotyped in old .config files but not new .config files:\n",
      "*****************************************************************\n",
      "3        (20, 1232)\n",
      "34       (20, 1393)\n",
      "45    (20, 2006174)\n",
      "47    (20, 2520669)\n",
      "Name: P/S Pair, dtype: object\n",
      "Length:\n",
      "4\n",
      "Pairs that were correctly genotyped in new .config files but not in old .config files:\n",
      "*****************************************************************\n",
      "18    (8, 2599208)\n",
      "Name: P/S Pair, dtype: object\n",
      "Length:\n",
      "1\n",
      "\n",
      "Pairs that were incorrectly genotyped in both analyses:\n",
      "*****************************************************************\n",
      "0          (4, 217)\n",
      "21        (4, 1232)\n",
      "22        (4, 1393)\n",
      "30       (4, 34966)\n",
      "1          (8, 217)\n",
      "23        (8, 1393)\n",
      "7         (8, 1791)\n",
      "25       (10, 1791)\n",
      "24       (14, 1393)\n",
      "31      (14, 34966)\n",
      "16    (14, 2520669)\n",
      "26       (15, 1791)\n",
      "28      (15, 34896)\n",
      "35    (15, 2520669)\n",
      "20        (17, 217)\n",
      "9       (18, 34896)\n",
      "14    (18, 2006174)\n",
      "2         (20, 217)\n",
      "27       (20, 1791)\n",
      "29      (20, 34896)\n",
      "32      (20, 34966)\n",
      "12    (20, 2006092)\n",
      "37    (20, 2599208)\n",
      "33      (23, 34966)\n",
      "3         (25, 217)\n",
      "4        (25, 1232)\n",
      "5        (25, 1393)\n",
      "8        (25, 1791)\n",
      "10      (25, 34896)\n",
      "11      (25, 34966)\n",
      "13    (25, 2006092)\n",
      "15    (25, 2006174)\n",
      "17    (25, 2520669)\n",
      "19    (25, 2599208)\n",
      "34      (26, 34966)\n",
      "38    (29, 2599208)\n",
      "6        (30, 1393)\n",
      "36    (30, 2520669)\n",
      "Name: P/S Pair, dtype: object\n",
      "Length:\n",
      "38\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load in and combine mismatch data\n",
    "full_mismatch_multiple_flanks_tsv = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/STRaitRazor/results/\" +\\\n",
    "                             \"alleles/full_mismatch_STRait_razor_alleles_0.15_multiple_flanks.tsv\",sep=\"\\t\")\n",
    "pd.options.display.max_colwidth = 100\n",
    "full_mismatch_multiple_flanks_tsv = full_mismatch_multiple_flanks_tsv.reindex(sorted \\\n",
    "                                        (full_mismatch_multiple_flanks_tsv.columns), axis=1)\n",
    "single_mismatch_multiple_flanks_tsv = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/STRaitRazor/results/\" +\\\n",
    "                             \"alleles/single_mismatch_STRait_razor_alleles_0.15_multiple_flanks.tsv\",sep=\"\\t\")\n",
    "pd.options.display.max_colwidth = 100\n",
    "single_mismatch_multiple_flanks_tsv = single_mismatch_multiple_flanks_tsv.reindex(sorted \\\n",
    "                                        (single_mismatch_multiple_flanks_tsv.columns), axis=1)\n",
    "\n",
    "all_mismatch_multiple_flanks_tsv = pd.concat((full_mismatch_multiple_flanks_tsv,single_mismatch_multiple_flanks_tsv),\n",
    "                                            ignore_index=True)\n",
    "all_mismatch_multiple_flanks_tsv[\"P/S Pair\"] = all_mismatch_multiple_flanks_tsv[[\"Primer\",\"Sample\"]].apply(tuple,axis=1)\n",
    "\n",
    "\n",
    "full_mismatch_single_flanks_tsv = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/STRaitRazor/results/\" +\\\n",
    "                             \"alleles/full_mismatch_STRait_razor_alleles_0.15_flank10.tsv\",sep=\"\\t\")\n",
    "pd.options.display.max_colwidth = 100\n",
    "full_mismatch_single_flanks_tsv = full_mismatch_single_flanks_tsv.reindex(sorted \\\n",
    "                                        (full_mismatch_single_flanks_tsv.columns), axis=1)\n",
    "single_mismatch_single_flanks_tsv = pd.read_csv(\"C:/Users/vik16/OneDrive/Documents/FOGS Summer 2022/STRaitRazor/results/\" +\\\n",
    "                             \"alleles/single_mismatch_STRait_razor_alleles_0.15_flank10.tsv\",sep=\"\\t\")\n",
    "pd.options.display.max_colwidth = 100\n",
    "single_mismatch_single_flanks_tsv = single_mismatch_single_flanks_tsv.reindex(sorted \\\n",
    "                                        (single_mismatch_single_flanks_tsv.columns), axis=1)\n",
    "\n",
    "all_mismatch_single_flanks_tsv = pd.concat((full_mismatch_single_flanks_tsv,single_mismatch_single_flanks_tsv),\n",
    "                                            ignore_index=True)\n",
    "all_mismatch_single_flanks_tsv[\"P/S Pair\"] = all_mismatch_single_flanks_tsv[[\"Primer\",\"Sample\"]].apply(tuple,axis=1)\n",
    "\n",
    "\n",
    "new_config_errors=all_mismatch_multiple_flanks_tsv[\"P/S Pair\"]\n",
    "old_config_errors=all_mismatch_single_flanks_tsv[\"P/S Pair\"]\n",
    "\n",
    "\n",
    "print(\"Number of pairs that didn't run with old .config files but ran with new .config files:\")\n",
    "print(\"*****************************************************************\")\n",
    "print(len(new_config_new_genotypes))\n",
    "print(\"\")\n",
    "\n",
    "\n",
    "print(\"Pairs that didn't run with old .config files and were incorrectly genotyped with new .config files:\")\n",
    "print(\"*****************************************************************\")\n",
    "new_config_not_run_before_errors=new_config_errors[new_config_errors.isin(single_flanks_failed_pairings)]\n",
    "print(new_config_not_run_before_errors)\n",
    "print(\"Length:\")\n",
    "print(len(new_config_not_run_before_errors))\n",
    "\n",
    "print(\"Pairs that didn't run with old .config files and were correctly genotyped with new .config files:\")\n",
    "print(\"*****************************************************************\")\n",
    "new_config_not_run_before_errors=new_config_new_genotypes[~new_config_new_genotypes.isin(new_config_errors)]\n",
    "print(new_config_not_run_before_errors)\n",
    "print(\"Length:\")\n",
    "print(len(new_config_not_run_before_errors))\n",
    "\n",
    "\n",
    "print(\"Pairs that were correctly genotyped in old .config files but not new .config files:\")\n",
    "print(\"*****************************************************************\")\n",
    "new_config_actual_errors=new_config_errors[~new_config_errors.isin(old_config_errors)]\n",
    "new_config_actual_errors=new_config_actual_errors[~new_config_actual_errors.isin(single_flanks_failed_pairings)]\n",
    "print(new_config_actual_errors.iloc[np.argsort(new_config_actual_errors)])\n",
    "print(\"Length:\")\n",
    "print(len(new_config_actual_errors))\n",
    "\n",
    "\n",
    "print(\"Pairs that were correctly genotyped in new .config files but not in old .config files:\")\n",
    "print(\"*****************************************************************\")\n",
    "old_but_not_new_errors=old_config_errors[~old_config_errors.isin(new_config_errors)]\n",
    "print(old_but_not_new_errors.iloc[np.argsort(old_but_not_new_errors)])\n",
    "print(\"Length:\")\n",
    "print(len(old_but_not_new_errors))\n",
    "print(\"\")\n",
    "\n",
    "print(\"Pairs that were incorrectly genotyped in both analyses:\")\n",
    "print(\"*****************************************************************\")\n",
    "shared_errors=old_but_not_new_errors=old_config_errors[old_config_errors.isin(new_config_errors)]\n",
    "print(shared_errors.iloc[np.argsort(shared_errors)])\n",
    "print(\"Length:\")\n",
    "print(len(shared_errors))\n",
    "print(\"\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nearestneighborsenv",
   "language": "python",
   "name": "nearestneighborsenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
